### PR TITLE
Expose List properties, remove Add* methods

### LINKED ISSUE #1884: Cannot replace default completions of Option<T> where T is enum
When `T` is an enum type, `Option<T>` gets a default completion source that returns the names of the defined enum values. If I call OptionExtensions.AddCompletions to add a completion callback that returns some of those names but not all, that has no effect because the default completion source returns all of them anyway and Argument.GetCompletions discards my completion items as duplicates, even if they contain documentation and the default completion items don't. Because the Option.Argument property is internal, I cannot call option.Argument.Completions.Clear() to delete the default completion source before I add mine.

Possibly I could hack around this by deriving my own class from `Option<T>` and overriding GetCompletions there.

## To reproduce

```XML
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.CommandLine" Version="2.0.0-beta4.22272.1" />
  </ItemGroup>

</Project>
```

```C#
namespace EnumCompletion;

using System.CommandLine;
using System.CommandLine.Completions;

class Program
{
    static void Main(string[] args)
    {
        var option = new Option<DayOfWeek>("-d")
            .AddCompletions(CompleteDayOfWeek);
        foreach (var completion in option.GetCompletions())
        {
            Console.WriteLine($"{completion.Label}: {completion.Documentation}");
        }
    }

    static IEnumerable<CompletionItem> CompleteDayOfWeek(CompletionContext context)
    {
        Console.Error.WriteLine("CompleteDayOfWeek was called.");

        // Omit Saturday and Sunday.
        return new[]
        {
           new CompletionItem("Monday", sortText: "1", documentation: "maanantai"),
           new CompletionItem("Tuesday", sortText: "2", documentation: "tiistai"),
           new CompletionItem("Wednesday", sortText: "3", documentation: "keskiviikko"),
           new CompletionItem("Thursday", sortText: "4", documentation: "torstai"),
           new CompletionItem("Friday", sortText: "5", documentation: "perjantai"),
        };
    }
}
```

## Actual output

```
CompleteDayOfWeek was called.
Friday:
Monday:
Saturday:
Sunday:
Thursday:
Tuesday:
Wednesday:
```

## Expected output

```
CompleteDayOfWeek was called.
Monday: maanantai
Tuesday: tiistai
Wednesday: keskiviikko
Thursday: torstai
Friday: perjantai
```

### LINKED ISSUE #1972: Symbols: Validators and Completions
Based on various recent discussions with @Jozkee and @KalleOlaviNiemitalo

https://github.com/dotnet/command-line-api/pull/1953#issuecomment-1318607961
https://github.com/dotnet/command-line-api/pull/1959/files#r1025309941

I've realized that:
- we currently don't expose a possibility to remove `Completions` from `Option<T>` (Argument exposes `.Completions` property, but  Option does not). Most of the methods exposed by Option simply delegate to `.Argument` which is an internal property.
- there is more than one way of adding new completions and validators: `symbol.AddValidator` and `symbol.Validators.Add`

The questions I have:

- should we simplify the public API surface by just exposing the mutable collections and removing `Add*` methods? @bartonjs what is the recommended approach for such scenarios? The change I am talking about:

```diff
public class Argument<T>
{
    public ICollection<Func<CompletionContext, IEnumerable<CompletionItem>>> Completions { get; }
-   public Argument<T> AddCompletions(Func<CompletionContext, IEnumerable<CompletionItem>> completionsDelegate)
// we don't expose ClearCompletions()
}
```

- should we remove the `Option` methods that just delegate to `Argument` and expose `Option.Argument`? @jonsequitur this has most likely been discussed in the past, what is the main reason for not exposing this property?

```diff
public class Option<T>
{
+   public Argument<T> { get; }
-    public void SetDefaultValue(T value);
-    public void SetDefaultValueFactory(Func<T> defaultValueFactory)
-    public Option<T> AcceptOnlyFromAmong(params string[] values)
-    public Option<T> AddCompletions(params string[] completions)
-    public Option<T> AddCompletions(Func<CompletionContext, IEnumerable<string>> completionsDelegate)
-    public Option<T> AddCompletions(Func<CompletionContext, IEnumerable<CompletionItem>> completionsDelegate)
-    public Option<T> AcceptLegalFilePathsOnly()
-    public Option<T> AcceptLegalFileNamesOnly()
}
```

It would simplify the public API surface and we would avoid the issues where not every method exposed by `Argument` is exposed by `Option`, but at a cost of losing the fluent API builder capabilities.

### PR DESCRIPTION
**Pull request by adamsitnik**
Based on feedback from @bartonjs

By exposing a single `List<T>` returning property we allow the users to not only add new items, but also remove them (#1884)

fixes #1972
fixes #1884

### CODE REVIEW

File: src/System.CommandLine.ApiCompatibility.Tests/ApiCompatibilityApprovalTests.System_CommandLine_api_is_not_changed.approved.txt [hunk 1]
```diff
@@ -1,9 +1,10 @@
 System.CommandLine
   public abstract class Argument : Symbol, System.CommandLine.Binding.IValueDescriptor
     public ArgumentArity Arity { get; set; }
-    public System.Collections.Generic.ICollection<System.Func<System.CommandLine.Completions.CompletionContext,System.Collections.Generic.IEnumerable<System.CommandLine.Completions.CompletionItem>>> Completions { get; }
+    public System.Collections.Generic.List<System.Func<System.CommandLine.Completions.CompletionContext,System.Collections.Generic.IEnumerable<System.CommandLine.Completions.CompletionItem>>> Completions { get; }
     public System.Boolean HasDefaultValue { get; }
     public System.String HelpName { get; set; }
+    public System.Collections.Generic.List<System.Action<System.CommandLine.Parsing.ArgumentResult>> Validators { get; }
     public System.Type ValueType { get; }
     public System.Collections.Generic.IEnumerable<System.CommandLine.Completions.CompletionItem> GetCompletions(System.CommandLine.Completions.CompletionContext context)
```
[the remark this answers is not in the document]
Comment (adamsitnik) [reply src/System.CommandLine.ApiCompatibility.Tests/ApiCompatibilityApprovalTests.System_CommandLine_api_is_not_changed.approved.txt:9]: @Jozkee that is a very good point!

@KathleenDollard what is your opinion on this?

File: src/System.CommandLine.ApiCompatibility.Tests/ApiCompatibilityApprovalTests.System_CommandLine_api_is_not_changed.approved.txt [hunk 2]
```diff
@@ -23,10 +24,8 @@ System.CommandLine
     public Argument<T> AcceptLegalFileNamesOnly()
     public Argument<T> AcceptLegalFilePathsOnly()
     public Argument<T> AcceptOnlyFromAmong(System.String[] values)
-    public Argument<T> AddCompletions(System.String[] completions)
     public Argument<T> AddCompletions(System.Func<System.CommandLine.Completions.CompletionContext,System.Collections.Generic.IEnumerable<System.String>> completionsDelegate)
     public Argument<T> AddCompletions(System.Func<System.CommandLine.Completions.CompletionContext,System.Collections.Generic.IEnumerable<System.CommandLine.Completions.CompletionItem>> completionsDelegate)
```
[the remark this answers is not in the document]
Comment (adamsitnik) [reply src/System.CommandLine.ApiCompatibility.Tests/ApiCompatibilityApprovalTests.System_CommandLine_api_is_not_changed.approved.txt]: @Jozkee great catch, #solved

File: src/System.CommandLine.ApiCompatibility.Tests/ApiCompatibilityApprovalTests.System_CommandLine_api_is_not_changed.approved.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamsitnik) [reply src/System.CommandLine.ApiCompatibility.Tests/ApiCompatibilityApprovalTests.System_CommandLine_api_is_not_changed.approved.txt:9]: Awesome, I've applied the rename and the PR is ready for review again. Thank you both for your feedback!

**Comment by adamsitnik** [comment]
I've addressed all the feedback so I am going to merge it to make some progress with other PRs.

File: src/System.CommandLine.Benchmarks/CommandLine/Perf_Parser_Directives_Suggest.cs [hunk 3]
```diff
@@ -22,10 +22,16 @@ public void Setup()
         {
             _nullConsole = new NullConsole();
 
+            Option<string> fruitOption = new("--fruit");
+            fruitOption.CompletionSources.Add("apple", "banana", "cherry");
```
Comment (jonsequitur) [inline src/System.CommandLine.Benchmarks/CommandLine/Perf_Parser_Directives_Suggest.cs:26]: This naming now looks wrong. The things being added here are really completion items, not completion sources.

Comment (adamsitnik) [reply src/System.CommandLine.Benchmarks/CommandLine/Perf_Parser_Directives_Suggest.cs:26]: @jonsequitur I've tried renaming the `Add` extension method to `AddCompletionItems` but it break the C# duck typing for collection initialization:

image

```log
Error	CS1503	Argument 1: cannot convert from 'string' to 'System.Func<System.CommandLine.Completions.CompletionContext, System.Collections.Generic.IEnumerable<System.CommandLine.Completions.CompletionItem>>'	System.CommandLine.Tests (net462), System.CommandLine.Tests (net7.0)	D:\projects\command-line-api\src\System.CommandLine.Tests\CompletionTests.cs	125	Active
```

**Pull request merged**